spatialos-sdk/src/worker/schema/component_data.rs

```diff
@@ -1,29 +1,66 @@
-use crate::worker::schema::SchemaObject;
+use crate::worker::schema::{owned::*, SchemaObject};
 use spatialos_sdk_sys::worker::*;
+use std::marker::PhantomData;
 
+/// Serialized schema data for a component, owned by the Rust SDK.
+///
+/// For maximum efficiency, the serialized data may borrow data from the component
+/// used to create an `OwnedComponentData` instance. The lifetime parameter
+/// tracks this borrow, such that an `OwnedComponentData` cannot outlive the
+/// data it borrows.
 #[derive(Debug)]
-pub struct SchemaComponentData {
-    pub(crate) internal: *mut Schema_ComponentData,
-}
+pub struct SchemaComponentData(PhantomData<*mut Schema_ComponentData>);
 
 impl SchemaComponentData {
-    pub fn new() -> SchemaComponentData {
-        SchemaComponentData {
-            internal: unsafe { Schema_CreateComponentData() },
-        }
+    pub fn new() -> Owned<SchemaComponentData> {
+        unsafe { Owned::new(Schema_CreateComponentData()) }
     }
 
     pub fn fields(&self) -> &SchemaObject {
-        unsafe { SchemaObject::from_raw(Schema_GetComponentDataFields(self.internal)) }
+        unsafe { SchemaObject::from_raw(Schema_GetComponentDataFields(self.as_ptr())) }
     }
 
     pub fn fields_mut(&mut self) -> &mut SchemaObject {
-        unsafe { SchemaObject::from_raw_mut(Schema_GetComponentDataFields(self.internal)) }
+        unsafe { SchemaObject::from_raw_mut(Schema_GetComponentDataFields(self.as_ptr())) }
+    }
+
+    // Methods for raw pointer conversion.
+    // -----------------------------------
+
+    pub(crate) unsafe fn from_raw<'a>(raw: *mut Schema_ComponentData) -> &'a Self {
+        &*(raw as *mut _)
+    }
+
+    pub(crate) fn as_ptr(&self) -> *mut Schema_ComponentData {
```

Comment:
Shouldn't this take `&mut self`?

Reply:
Not that it _really_ matters since you can copy the pointer, but it at least makes the intent explicit

Reply:
There are cases where we want to use a `&SchemaComponentData` with the C API, e.g. in the `fields` method. In an earlier iteration I had two methods for getting the raw poiner:

```
fn as_ptr(&self) -> *mut Schema_ComponentData { ... }

fn as_ptr_mut(&mut self) -> *mut Schema_ComponentData { ... }
```

Both have to return a `*mut` since that's what all of the C API functions take (rightfully so, since most of the methods perform some kind of internal mutation in certain cases), so we can't even use the difference to enforce that you need a `&mut` to get a `*mut`. I think it's ultimately better to just have one way to get the raw pointer out, and use other methods for documenting intent and enforcing safety (e.g. expose all meaningful operations as methods on `SchemaComponentData`, and avoid going through the raw C API outside of this module).